File: marketdata-common/src/main/java/com/handson/dto/Vwap.java
package com.handson.dto;

import com.google.gson.Gson;

public class Vwap {

    public final String code;
    public final double vwap;
    public final double volume;

    public Vwap(String code) {
        this.code = code;
        this.vwap = 0;
        this.volume = 0;
    }

    public Vwap(String code, double vwap, double volume) {
        this.code = code;
        this.vwap = vwap;
        this.volume = volume;
    }

    public Vwap addTrade(Trade t) {
        double volume = this.volume + t.quantity;
        double vwap = (this.volume * this.vwap + t.nominal) / volume;
        return new Vwap(t.code, vwap, volume);
    }

    public static Vwap fromJson(String input) {
        return new Gson().fromJson(input, Vwap.class);
    }


    public String toJson() {
        return new Gson().toJson(this);
    }

    @Override
    public String toString() {
        return toJson();
    }
}

File: CSharp/NetCore/MarketData/src/MarketDataCommon/Dto/Vwap.cs
using Newtonsoft.Json;

namespace MarketDataCommon.Dto
{
    public class Vwap
    {
        public string Code { get; set; }
        public double VwapValue { get; set; }
        public double Volume { get; set; }

        public Vwap(string code, double vwapValue, double volume)
        {
            Code = code;
            VwapValue = vwapValue;
            Volume = volume;
        }

        public Vwap AddTrade(Trade trade)
        {
            double volume = Volume + trade.Quantity;
            double vwapValue = (Volume*VwapValue + trade.Nominal)/Volume;
            return new Vwap(trade.Code, vwapValue, volume);
        }

        public static Vwap FromJson(string input)
        {
            return JsonConvert.DeserializeObject<Vwap>(input);
        }

        public string ToJson()
        {
            return JsonConvert.SerializeObject(this);
        }

        public override string ToString() => ToJson();

    }
}
